### DISCUSSION TITLE
What was special with version 3.4 (cg->version==3400)

### DISCUSSION THREAD

**Discussion by rsm73S** [CGNS/CGNS#355] [CGNStalk]
I have problems reading CGA file stored with version 3.4 (cg->version==3400). The section within the zone (only one zone within the only one base) gets no initialization of the connect_offset array (the pointer connect_offset remains 0x0) which leads to further segmentation fault inside cg_poly_elements_partial_read. I found very interesting if sentence

                if (cg->version < 4000 && cg->version != 3400) {

This is the line number 1400 in cgns_internals.c (inside function cgi_read_section). It seems to be my problem. So my version is exactly 3400. I.e. the whole block after line 1400 is not executed and the connect_offset array remains uninitialized.
My question to developers is: what was special with version 3400? Is it something like "a not able to work" version and do I have any chance to read the section information properly?

**Comment by rsm73S** [comment]
If I replace the line 1400 in cgns_internals.c with 

if (cg->version < 4000) {

anything works fine. 

So it seems to be a compatibility issue for unstructured meshes between the storing and the loading functions  at version 3400

**Comment by awoimbee** [comment]
What happened was that they included [CPEX 0041](https://cgns.github.io/ProposedExtensions/NGON-CPEX-0041-v0.16.pdf) in the v3.4.0 release.
Then they saw that it was a breaking change and released v3.4.1, which removes the move to CPEX 0041, but they fucked up and v3.4.1 saves files as version "3.4", just like v3.4.0.
So now a CGNS file version "3.4" might use the old or the new format for NFACE_n/NGON_n and MIXED connectivity. There is also a v3.4.2 I think but I don't know what it changed.

On top of that, even the CGNS MLL doesn't handle this case correctly. So we have to handle it ourselves.

https://cgnsorg.atlassian.net/browse/CGNS-285

**Comment by mennodeij** [comment]
v3.4.2 has the correct versioning in the file. Both 3.4.0 and 3.4.1 were retracted, i.e. "unreleased", because of the reason you mention @awoimbee . There are still programs out there that use 3.4.1 and when files written with that version are read by 4.0+, it tries to read the new format from an old-format file, resulting in incomplete reads. 
Files written with 3.4.2 are read correctly with 4.0+ as far as I know.

**Comment by rsm73S** [comment]
Thanx. This is also my experience. I still deal with 3.4.0 files which I unable to load with original code of 4.0+. So my recipe is to manipulate the cgns_internals.c, compile and work with it. And hope that some day I will get files stored with 4.0+

**Reply by DGPardo** [reply]
Hi @rsm73S 

During the release of version CGNS 3.4.1 they forgot to update the version numbering that is hardcoded in the library. Meaning that CGNS 3.4.1 was writing files with a new convention for NGON_n, MIXED and FACE but version indicator in the file still said it had the previous version.

This means that the CGNS code, by looking at the version label, cannot know if file is v3.4.0 or v3.4.1.

I think CGNS V4+ is only failing (non-backwards compatible) with files generated using version v3.4.1. So I am assuming that your files are actually 3.4.1 and not 3.4.0.

Having said so, CGNS v3.4.2 is what CGNS v3.4.1 should have been. As you can see in this diff, they are only fixing the version number https://github.com/CGNS/CGNS/compare/v3.4.1...v3.4.2

Rather than your complicated workaround changing `cgns_internals.c`, I would recommend you using cgnsview and modify the version number or write some sort of python script to do it (either with pyCgns in HDF5). Should be quite easy.

**Comment by smguzik** [comment]
To clarify, the inclusion of CPEX 41 in release 3.4.0 broke forwards compatibility.  Within a major release number, it is expected that earlier versions of the library can read files written by later versions (e.g. library 3.2 can read a file written by library 3.3).  There was no feasible way to "retract" a release so the fixes happened going forward:

1. Release 3.4.2 removed changes related to CPEX 41, but retained other updates.  Files created with 3.4.2 are forwards compatible within 3.x releases of the library.  They should also be backwards compatible with 4.x releases of the library.
2. Release 4.0.0 added CPEX 41 back in but the major version bump means files written by 4.x (possibly) cannot be read by version 3.x of the library.

More information can be found here: https://cgnsorg.atlassian.net/wiki/spaces/CGNS/pages/220463122/Resolve+issue+with+release+s+3.4.0+version+compatibility+the+4.0.0+release+and+forward+compatibility.
